File: problems/CSES/Graph_Algorithms/flight_routes_check/flight_routes_check.py
import sys
from collections import deque
# sys.setrecursionlimit(200000)

input = sys.stdin.readline

def bfs(start, graph):
    n = len(graph)
    visited = [False] * n

    queue = deque([start])
    visited[start] = True

    while queue:
        u = queue.popleft()

        for v in graph[u]:
            if not visited[v]:
                visited[v] = True
                queue.append(v)

    return visited

def main():
    n, m = map(int, input().split())

    graph = [[] for _ in range(n)]
    reverse_graph = [[] for _ in range(n)]

    for _ in range(m):
        a, b = map(int, input().split())
        a -= 1
        b -= 1

        graph[a].append(b)
        reverse_graph[b].append(a)

    visited = bfs(0, graph)

    for v in range(n):
        if not visited[v]:
            print("NO")
            print(1, v + 1)
            return

    visited_reverse = bfs(0, reverse_graph)

    for v in range(n):
        if not visited_reverse[v]:
            print("NO")
            print(v + 1, 1)
            return

    print("YES")

main()

File: problems/CSES/Graph_Algorithms/flight_routes_check/flight_routes_check.cpp
#include <bits/stdc++.h>
using namespace std;
using ll = long long;
#define pb push_back

vector<bool> bfs(int start, const vector<vector<int>>& graph) {
    int n = graph.size();

    vector<bool> visited(n, false);
    queue<int> q;

    q.push(start);
    visited[start] = true;

    while (!q.empty()) {
        int u = q.front();
        q.pop();

        for (int v : graph[u]) {
            if (!visited[v]) {
                visited[v] = true;
                q.push(v);
            }
        }
    }

    return visited;
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n, m;
    cin >> n >> m;

    vector<vector<int>> graph(n);
    vector<vector<int>> reverse_graph(n);

    for (int i = 0; i < m; i++) {
        int a, b;
        cin >> a >> b;

        a--;
        b--;

        graph[a].push_back(b);
        reverse_graph[b].push_back(a);
    }

    vector<bool> visited = bfs(0, graph);

    for (int v = 0; v < n; v++) {
        if (!visited[v]) {
            cout << "NO\n";
            cout << 1 << ' ' << v + 1 << '\n';
            return 0;
        }
    }

    vector<bool> visited_reverse = bfs(0, reverse_graph);

    for (int v = 0; v < n; v++) {
        if (!visited_reverse[v]) {
            cout << "NO\n";
            cout << v + 1 << ' ' << 1 << '\n';
            return 0;
        }
    }

    cout << "YES\n";

    return 0;
}
